### DISCUSSION TITLE
Are YubiKey's HMAC-SHA1 responses actually 20 bytes, or are they truncated?

### DISCUSSION THREAD

**Discussion by danimoh** [keepassxreboot/keepassxc#11608] [Q&A] [answered]
While the output length of SHA-1 is obviously 20 bytes, and that's also the length that [KeePassXC's code](https://github.com/keepassxreboot/keepassxc/blob/fb022cb5e964b702c8ea6b330d5b1878fb530504/src/keys/drivers/YubiKeyInterfaceUSB.cpp#L290) assumes being returned by YubiKey's HMAC-SHA1, I'm wondering whether that is actually the case? [Yubico's documentation](https://docs.yubico.com/yesdk/users-manual/application-otp/challenge-response.html#supported-challenge-response-algorithms:~:text=resulting%20in%20a%206%2D10%20digit%20HOTP%20as%20the%20response%20code) seems to hint at the response being truncated to 6-10 digits. When using a YubiKey as a second factor for unlocking the database, obviously 6-10 digits would add only limited additional protection over the master secret, and would be reasonable to brute force, if the master secret got leaked.
While I hope, and guess, that the documentation is simply inaccurate, wanted to open an issue for clarification nonetheless, as it would be quite bad, if the response is actually truncated.
Maybe someone who has more experience with such things than me can run a quick experiment that indeed 20 entirely different bytes are returned for different challenges, and not that the remainder is for example padded with 0s, or some other static bytes?

**Comment by droidmonkey** [comment, marked as answer]
Yes it is actually 20 bytes, see the Yubikey code here: 

https://github.com/keepassxreboot/keepassxc/blob/fb022cb5e964b702c8ea6b330d5b1878fb530504/src/thirdparty/ykcore/ykcore.c#L370-L407

I have no idea what the documentation is even saying by 6-10 digits, that makes zero sense


It's hard to tell because the response is literal bytes (not digits), but you can see the response is 20 bytes + residual which is then truncated to 20 bytes of data (the actual response):

![image](https://github.com/user-attachments/assets/c78ee8c7-a588-4e09-834d-f2dd9c0715da)

![image](https://github.com/user-attachments/assets/6e482445-d7dc-404d-8944-09623eb7023e)

**Reply by danimoh** [reply]
Hello @droidmonkey. Thank you so much for taking the time to confirm this :hearts:
That's reassuring.
